### ISSUE TITLE
Clean up some workarounds after upgrading minimum kernel version

### ISSUE THREAD

**Issue by lzutao** [rust-lang/rust#74519]
> Besides `O_CLOEXEC`, which other things would we be able to clean up?

Good question! There are 6 workarounds in `libstd` for older Linux versions (that I could find).
Increasing the minimum version to 2.6.32 (aka 3rd Kernel LTS, aka RHEL 6) would fix 5 of them.
Code links are inline:

- [ ] [`File::open_c`](https://github.com/rust-lang/rust/blob/8b94e9e9188b65df38a5f1ae723617dc2dfb3155/src/libstd/sys/unix/fs.rs#L469-L505) uses the `O_CLOEXEC` flag with [`open(2)`](http://man7.org/linux/man-pages/man2/open.2.html) (mentioned above, added in 2.6.23)
- [ ] [`FileDesc::duplicate`](https://github.com/rust-lang/rust/blob/8b94e9e9188b65df38a5f1ae723617dc2dfb3155/src/libstd/sys/unix/fd.rs#L230-L275) uses `F_DUPFD_CLOEXEC` with [`fcntl(2)`](http://man7.org/linux/man-pages/man2/fcntl.2.html) (added in 2.6.24, there is also the mention of a bug occuring on "some linux kernel at some point")
- [ ] [`pipe::anon_pipe`](https://github.com/rust-lang/rust/blob/8b94e9e9188b65df38a5f1ae723617dc2dfb3155/src/libstd/sys/unix/pipe.rs#L21-L49) uses [`pipe2(2)`](http://man7.org/linux/man-pages/man2/pipe2.2.html) to atomically set the `O_CLOEXEC` flag on the pipe fds (added in 2.6.27)
- [ ] [`Socket::new_raw`](https://github.com/rust-lang/rust/blob/8b94e9e9188b65df38a5f1ae723617dc2dfb3155/src/libstd/sys/unix/net.rs#L75-L86) uses `SOCK_CLOEXEC` with [`socket(2)`](http://man7.org/linux/man-pages/man2/socket.2.html) (added in 2.6.27)
- [ ] [`Socket::accept`](https://github.com/rust-lang/rust/blob/8b94e9e9188b65df38a5f1ae723617dc2dfb3155/src/libstd/sys/unix/net.rs#L194-L215) uses [`accept4(2)`](http://man7.org/linux/man-pages/man2/accept4.2.html) to permit use of `SOCK_CLOEXEC` (added in 2.6.28)
- ~[`fs::copy`](https://github.com/rust-lang/rust/blob/8b94e9e9188b65df38a5f1ae723617dc2dfb3155/src/libstd/sys/unix/fs.rs#L877-L955) uses [`copy_file_range(2)`](http://man7.org/linux/man-pages/man2/copy_file_range.2.html) (added in 4.5, not fixed by this proposal)~

As you can see, the workarounds fixed by this proposal all have a similar flavor.

_Originally posted by @josephlr in https://github.com/rust-lang/rust/issues/62516#issuecomment-515869417_

**Comment by lzutao** [comment]
Kernel updated in  #74163 .

**Comment by cuviper** [comment]
As noted, `copy_file_range` needs a much better kernel 4.5, so we'll still need that fallback. But all the `CLOEXEC` stuff should be ok to assume available now.

**Comment by RalfJung** [comment]
The new minimal version is 2.6.27 from what I can see, so "Socket::accept uses accept4(2) to permit use of SOCK_CLOEXEC" still needs a fallback, right?

**Comment by cuviper** [comment]
The builder is using 2.6.32 headers, so I would call that our new minimum.

**Comment by RalfJung** [comment]
Ah right; I am actually not sure where I got the 27 from...

**Comment by cuviper** [comment]
The title of #62516 suggested 2.6.27, but we went a little higher to match the actual distros in use.

**Comment by cuviper** [comment]
@alexcrichton I know this is ancient history, but do you recall the problem with `F_DUPFD_CLOEXEC`? You introduced that comment about "some kernel" in 3d28b8b98e6e4f55ef4ecd8babf0a050f48a3d11, and you were the last to update it in 0fff73b64a9658c9a3aa228488964d09e63115b4. If the problem kernels were before 2.6.32, then I'd like to clean that up now.

**Comment by cuviper** [comment]
Just digging in kernel history, I did find one hiccup in https://github.com/torvalds/linux/commit/121977187ca0a7f20b848530deb04cc56167769b, but that bug was introduced and resolved during the 3.7 development cycle, not in a release.

**Comment by alexcrichton** [comment]
IIRC we just mirrored what musl did at the time, which looks like it [still does](https://git.musl-libc.org/cgit/musl/tree/src/fcntl/fcntl.c):

```c
		int ret = __syscall(SYS_fcntl, fd, F_DUPFD_CLOEXEC, arg);
		if (ret != -EINVAL) {
			if (ret >= 0)
				__syscall(SYS_fcntl, ret, F_SETFD, FD_CLOEXEC);
			return __syscall_ret(ret);
		}
```

Looks like the original mailing list thread is gone and there's no comments in the musl source, so I don't recall what happened here.

I suspect it would be safe to remove those workarounds and just assume `F_DUPFD_CLOEXEC` works one the first try until someone says differently.
